### PR TITLE
Bump: AGP to 4.2.2

### PR DESCRIPTION
**Pull request by marandaneto**
## :scroll: Description
Bump: AGP to 4.2.2

## :bulb: Motivation and Context
be compatible with latest AGP version

## :green_heart: How did you test it?
CI runs tests for 4.2.2 too

## :pencil: Checklist

- [X] I reviewed the submitted code
- [X] I added tests to verify the changes
- [ ] I updated the docs if needed
- [X] No breaking changes

### CODE REVIEW

File: buildSrc/src/main/java/Dependencies.kt [hunk 1]
```diff
@@ -11,6 +11,6 @@ object LibsVersion {
 }
 
 object Libs {
-    val AGP = "com.android.tools.build:gradle:${BuildPluginsVersion.AGP}"
+    conval AGP = "com.android.tools.build:gradle:${BuildPluginsVersion.AGP}"
     val JUNIT = "junit:junit:${LibsVersion.JUNIT}"
```
Comment (marandaneto) [inline buildSrc/src/main/java/Dependencies.kt:15]:
```suggestion
    conval AGP = "com.android.tools.build:gradle:${BuildPluginsVersion.AGP}"
    conval val JUNIT = "junit:junit:${LibsVersion.JUNIT}"
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,6 +1,7 @@
 # Unreleased
 
 * Bump: sentry-cli to 1.64.1
+* Bump: AGP to 4.2.0
```
Comment (marandaneto) [inline CHANGELOG.md:4]:
```suggestion
* Bump: AGP to 4.2.0 (#106)
```

File: plugin-build/src/test/kotlin/io/sentry/android/gradle/SentryTaskProviderTest.kt [hunk 3]
```diff
@@ -149,9 +149,11 @@ class SentryTaskProviderTest {
         android.applicationVariants.all {
             if (it.name == "debug") {
                 assertEquals("assembleDebug", getAssembleTask(it).name)
-            } else {
-                assertEquals("assembleRelease", getAssembleTask(it).name)
             }
+// TODO: throwing Could not create task ':assembleRelease'.
+//            else {
+//                assertEquals("assembleRelease", getAssembleTask(it).name)
+//            }
```
Comment (marandaneto) [inline plugin-build/src/test/kotlin/io/sentry/android/gradle/SentryTaskProviderTest.kt:156]: @cortinico this is failing when I played with the AGP 4.2.0, any ideas?

**Review by philipphofmann — COMMENTED** [review]
Apart from the broken test and todo, LGTM.

File: plugin-build/src/test/kotlin/io/sentry/android/gradle/SentryTaskProviderTest.kt (hunk 3 above)
Comment (cortinico) [reply plugin-build/src/test/kotlin/io/sentry/android/gradle/SentryTaskProviderTest.kt:156]: I've done a quick test but I wasn't able to find out anything useful. Will have to look more into this.

**Comment by cortinico** [comment]
Small note: Bumping AGP to 4.2.0 will make the whole project unusable from IntelliJ IDEA 2021.1 (you could open it only with Android Studio). So unless this is urgent, I would hold off a bit.

**Comment by marandaneto** [comment]
> Small note: Bumping AGP to 4.2.0 will make the whole project unusable from IntelliJ IDEA 2021.1 (you could open it only with Android Studio). So unless this is urgent, I would hold off a bit.

its ok holding the upgrade, but we'll need to figure out this https://github.com/getsentry/sentry-android-gradle-plugin/pull/106#discussion_r628037240 because it could be that it's broken with AGP 4.2.0

**Comment by cortinico** [comment]
Seems like the failure is on Gradle/AGP side, specifically in the interaction with `ProjectBuilder`:

Gradle: https://github.com/gradle/gradle/issues/16774
AGP: https://issuetracker.google.com/issues/184334436

**Comment by cortinico** [comment]
I've rebased + worked on top of this PR @marandaneto
I believe that if CI is green we can merge this 🚀 

As a side note: I had to bump Gradle to 7.1 to let the tests pass successfully

**Comment by cortinico** [comment]
So the CI is failing with:

```
  Line 03: import org.gradle.configurationcache.extensions.serviceOf
                             ^ Unresolved reference: configurationcache

  Line 32:             serviceOf<ModuleRegistry>().getModule("gradle-tooling-api-builders")
                       ^ Unresolved reference: serviceOf
```
For Gradle 6.5 and 6.1.
How do we want to proceed? Potentially we could remove those two cases from the Publishing Dry Run matrix, as those versions are also tested with the Integration tests

**Comment by marandaneto** [comment]
> > So the CI is failing with:
> > ```
> >   Line 03: import org.gradle.configurationcache.extensions.serviceOf
> >                              ^ Unresolved reference: configurationcache
> > 
> >   Line 32:             serviceOf<ModuleRegistry>().getModule("gradle-tooling-api-builders")
> >                        ^ Unresolved reference: serviceOf
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > For Gradle 6.5 and 6.1.
> > How do we want to proceed? Potentially we could remove those two cases from the Publishing Dry Run matrix, as those versions are also tested with the Integration tests
> 
> yeah sounds like a good approach, but just for understanding, is the reason that Gradle 6.1-6.5 don't support `configurationcache` or what? I think it's fine to exclude those versions if it's because we are using a feature that is not supported on older versions (which is only required because AGP has a bug).

this workaround will potentially go away when we bump to AGP 7.x right since we'll require newer Gradle versions as well.

**Comment by cortinico** [comment]
> is the reason that Gradle 6.1-6.5 don't support `configurationcache` or what?

The reason is that I'm using the workaround mentioned in gradle/gradle#16774 to use `ProjectBuilder` on a AGP 4.2 test. That workaround relies on `configurationcache` to be applied.

This means that this project can't be opened/executed on Gradle 6.5. But the consumer of the plugin are not affected by the workaround (that's what we're testing in the integration test).

**Comment by marandaneto** [comment]
> > is the reason that Gradle 6.1-6.5 don't support `configurationcache` or what?
> 
> The reason is that I'm using the workaround mentioned in [gradle/gradle#16774](https://github.com/gradle/gradle/issues/16774) to use `ProjectBuilder` on a AGP 4.2 test. That workaround relies on `configurationcache` to be applied.
> 
> This means that this project can't be opened/executed on Gradle 6.5. But the consumer of the plugin are not affected by the workaround (that's what we're testing in the integration test).

got it, let's do your suggestion then:

> Potentially we could remove those two cases from the Publishing Dry Run matrix, as those versions are also tested with the Integration tests

**Pull request merged**